scripts/render-vm-startup-script.sh

```diff
@@ -0,0 +1,147 @@
+# Tesla GPU. It is designed to run on startup, provision the VM if necessary,
+# build 3D renderers (currently only Filament), collect screenshots and create
+# a PR on Github if there are any changes.
+#
+# In order to use this script as a startup script in your own VM, uncomment the
+# variables below and fill them in as appropriate, or else make sure the
+# environment variables are set before this script is run.
+##
+
+
+# GIT_USER_EMAIL=user@example.com
+# GIT_USER_NAME="User Name"
+# GITHUB_USER=example
+# GITHUB_ACCESS_TOKEN=your-github-access-token-here
+# MODEL_VIEWER_REPOSITORY=https://$GITHUB_USER:$GITHUB_ACCESS_TOKEN@github.com/googlewebcomponents/model-viewer.git
+# MODEL_VIEWER_BASE_BRANCH=master
+
+
+# Set up build toolchain apt repos
+sudo -E apt-add-repository -y "ppa:ubuntu-toolchain-r/test"
+wget -O - https://apt.llvm.org/llvm-snapshot.gpg.key|sudo apt-key add -
+echo "deb http://apt.llvm.org/xenial/ llvm-toolchain-xenial-7 main" | sudo tee -a ${TRAVIS_ROOT}/etc/apt/sources.list >/dev/null
+
+# Check for CUDA and try to install NVIDIA drivers
+if ! dpkg-query -W cuda-10-0; then
+ curl -O http://developer.download.nvidia.com/compute/cuda/repos/ubuntu1804/x86_64/cuda-repo-ubuntu1804_10.0.130-1_amd64.deb
+ sudo dpkg -i ./cuda-repo-ubuntu1804_10.0.130-1_amd64.deb
+ sudo apt-key adv --fetch-keys http://developer.download.nvidia.com/compute/cuda/repos/ubuntu1604/x86_64/7fa2af80.pub
+ sudo apt-get update
+ sudo apt-get install cuda-10-0 -y
+fi
+
+# Enable persistence mode
+sudo nvidia-smi -pm 1
+
+# Install build chain, graphics and X.org packages
+sudo apt install -y \
+  xserver-xorg \
+  freeglut3-dev \
+  mesa-common-dev \
+  mesa-utils \
+  libxmu-dev \
+  libglu1-mesa-dev \
+  libc++-7-dev \
+  libc++abi-7-dev \
+  libstdc++-7-dev \
+  clang-7 \
+  cmake \
+  ninja-build \
+  libxi-dev \
+  git
+
+# Start X.org as a daemon
+sudo /sbin/start-stop-daemon --start --quiet --pidfile /tmp/custom_xvfb_99.pid --make-pidfile --background --exec /usr/bin/X -- :0
+sleep 3 # Give X.org a few seconds to start up...
+
+# Export critical environment variables
+export DISPLAY=:0
+export CC=clang-7
+export CXX=clang++-7
+export FILAMENT_REQUIRES_CXXABI=true
+
+# Set up NVM and Node.js/NPM
+curl -o- https://raw.githubusercontent.com/creationix/nvm/v0.34.0/install.sh | bash
+export NVM_DIR="$HOME/.nvm"
+[ -s "$NVM_DIR/nvm.sh" ] && \. "$NVM_DIR/nvm.sh"
+nvm install --lts
+
+
+# Clone the repo if necessary
+if [ ! -d ./model-viewer ]; then
+  git clone $MODEL_VIEWER_REPOSITORY
+fi
+
+cd ./model-viewer
+
+# Update the repo from the origin and reset the tracked branch
+git fetch origin
+git add .
+git reset --hard origin/$MODEL_VIEWER_BASE_BRANCH
+
+# If Filament is still around, delete it so that we are sure to build it
+# cleanly from scratch
+if [ -d /tmp/filament ]; then
+  rm -rf /tmp/filament
+fi
+
+# Install NPM dependencies and update the screenshots
+npm install
+# NOTE: This will cause Filament to be compiled
+npm run update-screenshots
```

Comment:
Capturing the screenshots and running comparisons (which appears to be here down?) seems to have a different purpose than setting up the VM (above).

I might split them into two separate scripts and have an orchestration script to do them both? It won't be functionally different but I think it might be clearer.

Reply:
👍 Good call, will fix